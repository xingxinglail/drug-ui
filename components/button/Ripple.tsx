import { FC, createElement, useRef } from 'react';
import { scopedClassMaker } from '../_util';
import './style/ripple.scss';

const scopedClass = scopedClassMaker('drug-ripple');

interface RippleProps {
    center?: boolean;
}

const getRippleSize = (width: number, height: number, positionX: number, positionY: number) => {
    // 得到点击位置距离最远的点，计算出这两点之间的距离
    const square = (x: number) => x * x;
    const coordinates = [[0, 0], [width, 0], [0, height], [width, height]].map(coordinate => {
        return Math.sqrt(square(coordinate[0] - positionX) + square(coordinate[1] - positionY));
    });
    // 最长边即为半径
    const radius = Math.max.apply({}, coordinates);
    const size = radius * 2;
    const left = positionX - size / 2;
    const top = positionY - size / 2;
    return { size, left, top };
};

const Ripple: FC<RippleProps> = (props) => {
    const ref = useRef<HTMLElement>(null);
    let childDom: HTMLSpanElement;

    const handleMouseDown = (e: MouseEvent) => {
        if (ref.current) {
            const { left: _left, top: _top, width, height } = ref.current.getBoundingClientRect();
            const spanDom = document.createElement('span');
            spanDom.className = scopedClass('visible');
            let { size, left, top } = getRippleSize(width, height, e.pageX - _left, e.pageY - _top);
            if (props.center) {
                size = Math.max(width, height);
                left = (width - size) / 2;
                top = (height - size) / 2;
            }
            spanDom.style.width = `${ size }px`;
            spanDom.style.height = `${ size }px`;
            spanDom.style.top = `${ top }px`;
            spanDom.style.left = `${ left }px`;
            childDom = document.createElement('span');
            childDom.className = scopedClass('child');
            ;
            spanDom.appendChild(childDom);
            ref.current.appendChild(spanDom);
            const remove = () => {
                childDom.removeEventListener('animationend', remove);
                spanDom.remove();
            };
            childDom.addEventListener('animationend', remove);
        }
    };

    const handleMouseUp = () => {
        if (childDom) childDom.classList.add(scopedClass('child-leaving'));
    };

    return createElement('span', {
        className: scopedClass(),
        ref,
        onMouseDown: handleMouseDown,
        onMouseUp: handleMouseUp,
        onMouseLeave: handleMouseUp,
        onDragLeave: handleMouseUp
    });
};

Ripple.defaultProps = {
    center: false
};

export default Ripple;
