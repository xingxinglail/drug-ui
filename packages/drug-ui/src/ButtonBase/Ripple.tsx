import * as React from 'react';
import * as PropTypes from 'prop-types';
import { createUseStyles } from '@drug-ui/styles';
import { styles } from './Ripple.style';
import { Theme } from '../styles';

interface RippleProps {
    center?: boolean;
}

type RippleClassProps = 'root' | 'visible' | 'child' | 'childLeaving';

const name = 'Ripple';
const useStyles = createUseStyles<Theme, RippleClassProps>(styles, { name });

const getRippleSize = (width: number, height: number, positionX: number, positionY: number) => {
    // 得到点击位置距离最远的点，计算出这两点之间的距离
    const square = (x: number) => x * x;
    const coordinates = [[0, 0], [width, 0], [0, height], [width, height]].map(coordinate => {
        return Math.sqrt(square(coordinate[0] - positionX) + square(coordinate[1] - positionY));
    });
    // 最长边即为半径
    const radius = Math.max.apply({}, coordinates);
    const size = radius * 2;
    const left = positionX - size / 2;
    const top = positionY - size / 2;
    return { size, left, top };
};

const Ripple: React.FC<RippleProps> = (props) => {
    const ref = React.useRef<HTMLElement>(null);
    let childDom: HTMLSpanElement;
    const classes = useStyles();
    const handleMouseDown = (e: MouseEvent) => {
        const root = ref.current;
        if (root) {
            const { left: _left, top: _top, width, height } = root.getBoundingClientRect();
            const spanDom = document.createElement('span');
            spanDom.className = classes.visible;
            let { size, left, top } = getRippleSize(width, height, e.pageX - _left, e.pageY - _top);
            if (props.center) {
                size = Math.max(width, height);
                left = (width - size) / 2;
                top = (height - size) / 2;
            }
            spanDom.style.width = `${ size }px`;
            spanDom.style.height = `${ size }px`;
            spanDom.style.top = `${ top }px`;
            spanDom.style.left = `${ left }px`;
            childDom = document.createElement('span');
            childDom.className = classes.child;
            spanDom.appendChild(childDom);
            root.appendChild(spanDom);
            const remove = () => {
                childDom.removeEventListener('animationend', remove);
                spanDom.remove();
            };
            childDom.addEventListener('animationend', remove);
        }
    };

    const handleMouseUp = () => {
        if (childDom) childDom.classList.add(classes.childLeaving);
    };

    return React.createElement('span', {
        className: classes.root,
        ref,
        onMouseDown: handleMouseDown,
        onMouseUp: handleMouseUp,
        onMouseLeave: handleMouseUp,
        onDragLeave: handleMouseUp
    });
};

Ripple.defaultProps = {
    center: false
};

Ripple.propTypes = {
    center: PropTypes.bool
};

export default Ripple;
