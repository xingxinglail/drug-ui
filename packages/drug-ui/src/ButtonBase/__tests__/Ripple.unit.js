import * as renderer from 'react-test-renderer';
import React from 'react';
import { mount } from 'enzyme';
import Ripple from '../Ripple';

describe('<Ripple />', () => {

    it('render successful', () => {
        const json = renderer.create(<Ripple />).toJSON();
        expect(json).toMatchSnapshot();
    });

    it('set props center', () => {
        const wrapper = mount(<Ripple />);
        expect(wrapper.props().center).toBeFalsy();
        wrapper.setProps({ center: true });
        expect(wrapper.props().center).toBeTruthy();
    });
});
