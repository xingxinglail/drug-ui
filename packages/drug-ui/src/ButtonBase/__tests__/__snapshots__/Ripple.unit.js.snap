// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ripple /> render successful 1`] = `
<span
  className="Ripple-root-0-2-1"
  onDragLeave={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
/>
`;
